### PR TITLE
fix #6472 proper truncate directory names

### LINKED ISSUE #6472: Dialog for dir selection truncated
Thats how it looks on my Samsung S4.
You are not able to see the free space on the target dirs:
screenshot_2017-04-03-21-00-12

### PR DESCRIPTION
**Pull request by pstorch**

### CODE REVIEW

**Comment by pstorch** [comment]
I don't think the codacy findings are valid here.
In the UnitTest I exactly wanted to test the sdcard names and the other finding is at a place I haven't changed.

File: main/src/cgeo/geocaching/utils/Formatter.java [hunk 1]
```diff
@@ -324,5 +324,31 @@ public static String formatBytes(final long bytes) {
...
+                if (!directories.get(i).toString().endsWith(newEnding)) {
+                    break LOOP;
+                }
+            }
+            commonEnding = newEnding;
+            offset = commonEnding.length() + 1;
+        }
+        final List<CharSequence> truncatedDirs = new ArrayList<>(directories.size());
+        for (int i = 0; i < directories.size(); i++) {
+            final CharSequence title = directories.get(i);
```
Comment (UniQP) [inline main/src/cgeo/geocaching/utils/Formatter.java]: You can use an enhanced for-loop here.

File: main/src/cgeo/geocaching/utils/Formatter.java [hunk 2]
```diff
@@ -324,5 +324,31 @@ public static String formatBytes(final long bytes) {
         return String.format(Locale.getDefault(), "%.1f %sB", bytes / Math.pow(1024, exp), pre);
     }
 
+    public static List<CharSequence> truncateCommonSubdir(final List<CharSequence> directories) {
```
Comment (UniQP) [inline main/src/cgeo/geocaching/utils/Formatter.java]: Do we have any guarantees on the length of `directories`? With length 0 we get an OutOfBoundsException and with length 1 the while loop will never terminate.

Comment (pstorch) [inline main/src/cgeo/geocaching/utils/Formatter.java]: This dialog is only shown if the list has a size > 1.

File: main/src/cgeo/geocaching/utils/Formatter.java (hunk 1 above)
Comment (pstorch) [inline main/src/cgeo/geocaching/utils/Formatter.java]: Thanks for pointing it out. This was not possible before my last refactoring.

File: main/src/cgeo/geocaching/utils/Formatter.java (hunk 2 above)
Comment (pstorch) [inline main/src/cgeo/geocaching/utils/Formatter.java]: I'll add a check to be on the safe side

File: main/src/cgeo/geocaching/utils/Formatter.java [hunk 3]
```diff
@@ -324,5 +324,33 @@ public static String formatBytes(final long bytes) {
...
+                    break LOOP;
+                }
+            }
+            commonEnding = newEnding;
+            offset = commonEnding.length() + 1;
+        }
+        final List<CharSequence> truncatedDirs = new ArrayList<>(directories.size());
+        for (final CharSequence title : directories) {
+            CharSequence truncated = title;
+            if (commonEnding != null) {
```
Comment (UniQP) [inline main/src/cgeo/geocaching/utils/Formatter.java]: The condition is loop invariant, so we can test it before the loop and simply return `directories` if `commonEnding` is `null`.

File: main/src/cgeo/geocaching/utils/Formatter.java [hunk 4]
```diff
@@ -324,5 +324,33 @@ public static String formatBytes(final long bytes) {
...
 
+    public static List<CharSequence> truncateCommonSubdir(@NonNull final List<CharSequence> directories) {
+        if (directories.size() < 2) {
+            return directories;
+        }
+        final String firstDir = directories.get(0).toString();
+        String commonEnding = null;
+        int offset = 0;
+        LOOP:
+        while (true) {
```
Comment (UniQP) [inline main/src/cgeo/geocaching/utils/Formatter.java]: I'm still not happy with this loop. I will try to rewrite it and report back.

Comment (pstorch) [inline main/src/cgeo/geocaching/utils/Formatter.java]: I'm not happy either. Ideas are welcome. 
I've written a Unit test for it, so refactorings can be verified easily. 
Thanks for your effort.

**Comment by pstorch** [comment]
Thanks @UniQP for simplifying the loop. I've just added another testcase and inlined a temp variable.

**Pull request merged**